scala/lambdas/ingest-flow-control/src/main/scala/uk/gov/nationalarchives/ingestflowcontrol/Lambda.scala

```diff
@@ -0,0 +1,210 @@
+            val partKeys = List(IngestQueuePartitionKey(currentSystem.systemName))
+            dependencies.dynamoClient
+              .getItems[IngestQueueTableItem, IngestQueuePartitionKey](partKeys, config.flowControlQueueTableName)
+              .flatMap { queueTableItems =>
+                val taskToken = queueTableItems.headOption.map(_.taskToken)
+                val queuedAt = queueTableItems.headOption.map(_.queuedAt)
+                if (taskToken.nonEmpty) {
+                  (dependencies.stepFunctionClient.sendTaskSuccess(taskToken.get) >>
+                    dependencies.dynamoClient.deleteItems(
+                      config.flowControlQueueTableName,
+                      List(IngestQueuePrimaryKey(IngestQueuePartitionKey(currentSystem.systemName), IngestQueueSortKey(queuedAt.get)))
+                    )) >>
+                    IO(true)
+                } else {
+                  startTaskOnDedicatedChannel(sourceSystems.tail, executionsMap, flowControlConfig, taskStarted)
+                }
+              }
+      }
+
+      /** A recursive method to send success to one of the systems based on probability as configured in the Flow control configuration. It builds a probability ranges map based on
+        * the config (e.g. TDR -> (1, 25), FCL -> (25, 40) ... ). It then generates a random number between the minimum and maximum value over all probabilities and tries to
+        * schedule a task for that system, If there is no task waiting for the system, it recreates the probability map excluding that system from the config and generates a random
+        * number for the remaining systems only, thus making sure that the probabilities are honoured over all iterations.
+        *
+        *  The probability of each system is kept intact in relation to each other, even if one of the systems does not have a waiting task.
+        *  e.g. if there are 3 systems, one, two, and three with probabilities of 25, 55, 20, the ranges being 1-26, 26-81, 81-101
+        *  Iteration 1 - random number generated is 30 (which corresponds to system 2), in case system "two" does not have a waiting task,
+        *  Iteration 2 - a new ranges map is constructed by excluding system "two" (1-26, 26-47) and a new number is generated between 1 and 47
+        *                this ensures that the probability in relation to each other is kept intact for the remaining systems
+        * @param sourceSystems
+        *   List of source systems to iterate for starting a task
+        * @return
+        *   IO[Unit]
+        */
+      def startTaskBasedOnProbability(sourceSystems: List[SourceSystem]): IO[Unit] = {
+        sourceSystems match
+          case Nil => IO.unit // nothing more to do, no one is waiting
+          case _ =>
+          val sourceSystemProbabilityMap = buildProbabilityRangesMap(sourceSystems, 1, Map.empty[String, (Int, Int)])
+          val maxRandomValue: Int = sourceSystemProbabilityMap.values.maxBy(_._2)._2
```

Comment:
Maybe `sourceSystemProbabilityMap.values.map(_._2).max` might be slightly clearer?